### PR TITLE
 add tracer spans SRX-G6KFIX

### PR DESCRIPTION
**Pull request by medo-freiheit**

### CODE REVIEW

File: services/cd-service/pkg/cmd/server.go [hunk 1]
```diff
@@ -91,6 +91,11 @@ func RunServer() {
 			defer tracer.Stop()
 		}
 
+		span, ok := tracer.SpanFromContext(ctx)
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/cmd/server.go:94]: is it ok to run this when `EnableTracing=false`?

File: services/cd-service/pkg/repository/repository.go [hunk 2]
```diff
@@ -369,10 +370,21 @@ func (r *repository) Apply(ctx context.Context, transformers ...Transformer) err
 }
 
 func (r *repository) Push(ctx context.Context, pushAction func() error) error {
+
+	span, ok := tracer.SpanFromContext(ctx)
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/repository.go:374]: I think `StartSpanFromContext` is actually the better option here, as it creates a new Span, if none exists.

Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/repository.go:374]:
```suggestion
	span, ctx := tracer.StartSpanFromContext(ctx, "Push")
```
Note that it returns a context which we need to use in the rest of the function

File: services/cd-service/pkg/cmd/server.go (hunk 1 above)
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/cmd/server.go:94]: I think we need to use StartSpanFromContext here as well

Comment (medo-freiheit) [reply services/cd-service/pkg/cmd/server.go]: According to the documentation: `If the tracer is not started, calling this function is a no-op.`

File: services/cd-service/pkg/repository/repository.go [hunk 3]
```diff
@@ -369,10 +370,16 @@ func (r *repository) Apply(ctx context.Context, transformers ...Transformer) err
...
 func (r *repository) Push(ctx context.Context, pushAction func() error) error {
+
+	span, _ := tracer.StartSpanFromContext(ctx, "Apply")
+	defer span.Finish()
+
 	eb := backoff.NewExponentialBackOff()
 	eb.MaxElapsedTime = 7 * time.Second
 	return backoff.Retry(
 		func() error {
+			span, _ := tracer.StartSpanFromContext(ctx, fmt.Sprintf("try to Push at: %v", eb.GetElapsedTime()))
```
Comment (hannesg) [inline services/cd-service/pkg/repository/repository.go:381]: The action should never be dynamic. If you want to add dynamic data to the span, please use `SetTag` like this:

```
span, _ := tracer.StartSpanFromContext(ctx, "push")
span.SetTag("elapsedTime", eb.GetElapsedTime())
```

File: services/cd-service/pkg/cmd/server.go [hunk 4]
```diff
@@ -91,6 +91,8 @@ func RunServer() {
 			defer tracer.Stop()
 		}
 
+		span, _ := tracer.StartSpanFromContext(ctx, "Start server")
```
Comment (hannesg) [inline services/cd-service/pkg/cmd/server.go:94]: If you discard the ctx return value, then your span will be propagated through the stack. Please use
```
span, ctx := ...
```

**Review by hannesg — CHANGES_REQUESTED** [review]

**Review by sven-urbanski-freiheit-com — APPROVED** [review]

**Review by hannesg — APPROVED** [review]

**Pull request merged**